File: src/routes/question/question.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  Query,
  HttpCode,
  HttpStatus,
  UseGuards,
  ParseIntPipe,
  UseInterceptors,
  UploadedFiles,
} from '@nestjs/common'
import { QuestionService } from './question.service'
import {
  CreateQuestionDTO,
  UpdateQuestionDTO,
  QueryQuestionDTO,
  BulkCreateQuestionsDTO,
  CloneQuestionDTO,
} from './question.dto'
import { Auth } from 'src/shared/decorators/auth.decorator'
import { AuthType } from 'src/shared/constants/auth.constant'
import { Roles } from 'src/shared/decorators/roles.decorator'
import { RolesGuard } from 'src/shared/guards/roles.guard'
import { RoleName } from 'src/shared/constants/role.constant'
import type { JLPTLevelType, QuestionTypeType, DifficultyType } from 'src/shared/constants/enum.constant'
import { FileFieldsInterceptor } from '@nestjs/platform-express'
import { imageUploadOptions, mediaUploadOptions } from 'src/shared/config/upload.config'

@Controller('questions')
@UseGuards(RolesGuard)
export class QuestionController {
  constructor(private readonly questionService: QuestionService) {}

  @Post()
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.CREATED)
  @UseInterceptors(
    FileFieldsInterceptor(
      [
        { name: 'image', maxCount: 1 },
        { name: 'audio', maxCount: 1 },
      ],
      mediaUploadOptions,
    ),
  )
  async create(
    @Body() createDto: CreateQuestionDTO,
    @UploadedFiles() files?: { image?: Express.Multer.File[]; audio?: Express.Multer.File[] },
  ) {
    return this.questionService.create(createDto, files)
  }

  @Post('bulk')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin)
  @HttpCode(HttpStatus.CREATED)
  async bulkCreate(@Body() bulkCreateDto: BulkCreateQuestionsDTO) {
    return this.questionService.bulkCreate(bulkCreateDto)
  }

  @Get()
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async findAll(@Query() queryDto: QueryQuestionDTO) {
    return this.questionService.findAll(queryDto)
  }

  @Get('statistics')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async getStatistics() {
    return this.questionService.getStatistics()
  }

  @Get('type/:type')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async findByType(@Param('type') type: QuestionTypeType, @Query() queryDto: Omit<QueryQuestionDTO, 'type'>) {
    return this.questionService.findByType(type, queryDto)
  }

  @Get('level/:level')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async findByLevel(@Param('level') level: JLPTLevelType, @Query() queryDto: Omit<QueryQuestionDTO, 'level'>) {
    return this.questionService.findByLevel(level, queryDto)
  }

  @Get('difficulty/:difficulty')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async findByDifficulty(
    @Param('difficulty') difficulty: DifficultyType,
    @Query() queryDto: Omit<QueryQuestionDTO, 'difficulty'>,
  ) {
    return this.questionService.findByDifficulty(difficulty, queryDto)
  }

  @Get(':id')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async findOne(@Param('id', ParseIntPipe) id: number) {
    return this.questionService.findOne(id)
  }

  @Put(':id')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  @UseInterceptors(
    FileFieldsInterceptor(
      [
        { name: 'image', maxCount: 1 },
        { name: 'audio', maxCount: 1 },
      ],
      mediaUploadOptions,
    ),
  )
  async update(
    @Param('id', ParseIntPipe) id: number,
    @Body() updateDto: UpdateQuestionDTO,
    @UploadedFiles() files?: { image?: Express.Multer.File[]; audio?: Express.Multer.File[] },
  ) {
    return this.questionService.update(id, updateDto, files)
  }

  @Delete(':id')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin)
  @HttpCode(HttpStatus.OK)
  async remove(@Param('id', ParseIntPipe) id: number) {
    await this.questionService.remove(id)
    return { message: 'Question deleted successfully' }
  }

  @Post(':id/clone')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.CREATED)
  async cloneQuestion(@Param('id', ParseIntPipe) id: number, @Body() cloneDto: CloneQuestionDTO) {
    return this.questionService.cloneQuestion(id, cloneDto)
  }

  @Get('uuid/:uuid/versions')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async getVersions(@Param('uuid') uuid: string) {
    return this.questionService.getQuestionVersions(uuid)
  }

  @Get('uuid/:uuid/version/:version')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async getVersion(@Param('uuid') uuid: string, @Param('version', ParseIntPipe) version: number) {
    return this.questionService.getQuestionByVersion(uuid, version)
  }

  @Get(':id/usage')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async checkUsage(@Param('id', ParseIntPipe) id: number) {
    return this.questionService.checkQuestionUsage(id)
  }
}

File: assessment-service/src/main/java/com/torii/assessment/controller/QuestionController.java
package com.torii.assessment.controller;

import com.torii.assessment.dto.question.*;
import com.torii.assessment.entity.Question;
import com.torii.assessment.service.QuestionService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/questions")
@RequiredArgsConstructor
@Tag(name = "Questions", description = "Question management APIs")
public class QuestionController {

    private final QuestionService questionService;

    @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    @Operation(summary = "Create a new question with optional image and audio files")
    public ResponseEntity<QuestionResponseDTO> createQuestion(
            @ModelAttribute @Valid CreateQuestionDTO dto,
            @RequestPart(value = "image", required = false) MultipartFile image,
            @RequestPart(value = "audio", required = false) MultipartFile audio) {
        QuestionResponseDTO question = questionService.createQuestion(dto, image, audio);
        return ResponseEntity.status(HttpStatus.CREATED).body(question);
    }

    @PostMapping("/bulk")
    @Operation(summary = "Bulk create questions")
    public ResponseEntity<List<QuestionResponseDTO>> bulkCreateQuestions(@Valid @RequestBody BulkCreateQuestionsDTO dto) {
        List<QuestionResponseDTO> questions = questionService.bulkCreateQuestions(dto);
        return ResponseEntity.status(HttpStatus.CREATED).body(questions);
    }

    @GetMapping
    @Operation(summary = "Get all questions with pagination and filters")
    public ResponseEntity<Map<String, Object>> getAllQuestions(
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer limit,
            @RequestParam(required = false) Question.QuestionType type,
            @RequestParam(required = false) Question.JLPTLevel level,
            @RequestParam(required = false) Question.Difficulty difficulty,
            @RequestParam(required = false) Question.ReadingLength readingLength,
            @RequestParam(required = false) String keyword,
            @RequestParam(required = false) Boolean hasMedia,
            @RequestParam(required = false, defaultValue = "createdAt") String sortBy,
            @RequestParam(required = false, defaultValue = "desc") String sortOrder) {

        QueryQuestionDTO queryDto = QueryQuestionDTO.builder()
            .page(page != null ? page : 1)
            .limit(limit != null ? limit : 10)
            .type(type)
            .level(level)
            .difficulty(difficulty)
            .readingLength(readingLength)
            .keyword(keyword)
            .hasMedia(hasMedia)
            .sortBy(sortBy)
            .sortOrder(sortOrder)
            .build();

        Map<String, Object> result = questionService.getAllQuestions(queryDto);
        return ResponseEntity.ok(result);
    }

    @GetMapping("/statistics")
    @Operation(summary = "Get question statistics")
    public ResponseEntity<QuestionStatsDTO> getStatistics() {
        QuestionStatsDTO stats = questionService.getStatistics();
        return ResponseEntity.ok(stats);
    }

    @GetMapping("/type/{type}")
    @Operation(summary = "Get questions by type")
    public ResponseEntity<Map<String, Object>> getQuestionsByType(
            @PathVariable Question.QuestionType type,
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer limit,
            @RequestParam(required = false, defaultValue = "createdAt") String sortBy,
            @RequestParam(required = false, defaultValue = "desc") String sortOrder) {

        QueryQuestionDTO queryDto = QueryQuestionDTO.builder()
            .page(page != null ? page : 1)
            .limit(limit != null ? limit : 10)
            .sortBy(sortBy)
            .sortOrder(sortOrder)
            .build();

        Map<String, Object> result = questionService.findByType(type, queryDto);
        return ResponseEntity.ok(result);
    }

    @GetMapping("/level/{level}")
    @Operation(summary = "Get questions by JLPT level")
    public ResponseEntity<Map<String, Object>> getQuestionsByLevel(
            @PathVariable Question.JLPTLevel level,
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer limit,
            @RequestParam(required = false, defaultValue = "createdAt") String sortBy,
            @RequestParam(required = false, defaultValue = "desc") String sortOrder) {

        QueryQuestionDTO queryDto = QueryQuestionDTO.builder()
            .page(page != null ? page : 1)
            .limit(limit != null ? limit : 10)
            .sortBy(sortBy)
            .sortOrder(sortOrder)
            .build();

        Map<String, Object> result = questionService.findByLevel(level, queryDto);
        return ResponseEntity.ok(result);
    }

    @GetMapping("/difficulty/{difficulty}")
    @Operation(summary = "Get questions by difficulty")
    public ResponseEntity<Map<String, Object>> getQuestionsByDifficulty(
            @PathVariable Question.Difficulty difficulty,
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer limit,
            @RequestParam(required = false, defaultValue = "createdAt") String sortBy,
            @RequestParam(required = false, defaultValue = "desc") String sortOrder) {

        QueryQuestionDTO queryDto = QueryQuestionDTO.builder()
            .page(page != null ? page : 1)
            .limit(limit != null ? limit : 10)
            .sortBy(sortBy)
            .sortOrder(sortOrder)
            .build();

        Map<String, Object> result = questionService.findByDifficulty(difficulty, queryDto);
        return ResponseEntity.ok(result);
    }

    @GetMapping("/{id}")
    @Operation(summary = "Get question by ID")
    public ResponseEntity<QuestionResponseDTO> getQuestionById(@PathVariable Long id) {
        QuestionResponseDTO question = questionService.getQuestionById(id);
        return ResponseEntity.ok(question);
    }

    @PutMapping(value = "/{id}", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    @Operation(summary = "Update a question with optional image and audio files")
    public ResponseEntity<QuestionResponseDTO> updateQuestion(
            @PathVariable Long id,
            @ModelAttribute @Valid UpdateQuestionDTO dto,
            @RequestPart(value = "image", required = false) MultipartFile image,
            @RequestPart(value = "audio", required = false) MultipartFile audio) {
        QuestionResponseDTO question = questionService.updateQuestion(id, dto, image, audio);
        return ResponseEntity.ok(question);
    }

    @DeleteMapping("/{id}")
    @Operation(summary = "Delete a question")
    public ResponseEntity<Map<String, String>> deleteQuestion(@PathVariable Long id) {
        questionService.deleteQuestion(id);
        return ResponseEntity.ok(Map.of("message", "Question deleted successfully"));
    }

    @PostMapping("/{id}/clone")
    @Operation(summary = "Clone a question to create a new version")
    public ResponseEntity<QuestionResponseDTO> cloneQuestion(
            @PathVariable Long id,
            @RequestBody(required = false) UpdateQuestionDTO modifications) {
        QuestionResponseDTO question = questionService.cloneQuestion(id,
            modifications != null ? modifications : new UpdateQuestionDTO());
        return ResponseEntity.status(HttpStatus.CREATED).body(question);
    }

    @GetMapping("/uuid/{uuid}/versions")
    @Operation(summary = "Get all versions of a question by UUID")
    public ResponseEntity<List<QuestionResponseDTO>> getQuestionVersions(@PathVariable String uuid) {
        List<QuestionResponseDTO> versions = questionService.getQuestionVersions(uuid);
        return ResponseEntity.ok(versions);
    }

    @GetMapping("/uuid/{uuid}/version/{version}")
    @Operation(summary = "Get a specific version of a question")
    public ResponseEntity<QuestionResponseDTO> getQuestionByVersion(
            @PathVariable String uuid,
            @PathVariable Integer version) {
        QuestionResponseDTO question = questionService.getQuestionByVersion(uuid, version);
        return ResponseEntity.ok(question);
    }

    @GetMapping("/{id}/usage")
    @Operation(summary = "Check if question is used in assessments or quizzes")
    public ResponseEntity<Map<String, Object>> checkQuestionUsage(@PathVariable Long id) {
        Map<String, Object> usage = questionService.checkQuestionUsage(id);
        return ResponseEntity.ok(usage);
    }
}

